### PR TITLE
move device to training

### PR DESCRIPTION
**Pull request by omkar-334**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Addresses #1442 

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by omkar-334** [comment]
The lint tests are failing because now we're importing the function `get_world_size_and_rank` two times, one from `torchtune.training._device` and the other from `torchtune.training._distributed`, which has a deprecated warning.   
```python
@deprecated(
    msg="`get_world_size_and_rank` will move to `torchtune.utils._device` in future releases. "
    "Please use `torchtune.utils.get_world_size_and_rank` instead."
)
```   
What would be the solution for this?

**Comment by omkar-334** [comment]
@pbontrager @RdoubleA can you review my PR ? Thanks!

**Review by joecummings — COMMENTED** [review]
Thanks so much for the contribution @omkar-334!

Unfortunately, I messed up :/ While I was going through and finding issues that would be good for first time contributors to take on, I found this one quickly updated it without thinking. I believe the thinking originally was that specifically the ``batch_to_device`` function did not belong in utils since it's only a training utility. After discussing with @pbontrager , the proper fix for something like that is *not* to just move device.py to the dir `training/` but rather use something like `tensordicts` to easily move data around on devices with the standard PyTorch `to` function. This will require a more in-depth piece of work and I'll write up an Issue at a later time for it.

We really do value your time and contributions, so I'm very sorry this one didn't work out! It looks like you're interested in contributing partially to #2715, which would also be a huge help.

**Comment by omkar-334** [comment]
> After discussing with @pbontrager , the proper fix for something like that is _not_ to just move device.py to the dir `training/` but rather use something like `tensordicts` to easily move data around on devices with the standard PyTorch `to` function. This will require a more in-depth piece of work and I'll write up an Issue at a later time for it.

Understood... No worries. hopefully I can contribute more and have a good understanding by then.  
Closing this PR.  

> It looks like you're interested in contributing partially to #2715, which would also be a huge help.  

Yup, I'll get started with it.   
Thanks!
